modules/pool/lib/staking/gauge-staking.service.ts

```diff
@@ -1,221 +1,252 @@
     }
 
-    async getChildChainGaugeInfo(gaugeAddresses: string[]): Promise<{ [gaugeAddress: string]: ChildChainInfo }> {
+    private async getOnchainRewardTokensData(gauges: { id: string, version: 1 | 2, tokens: { id: string }[] }[]) {
+        // Get onchain data for BAL rewards
         const currentWeek = Math.floor(Date.now() / 1000 / 604800);
-        const childChainAbi =
-            networkContext.chain === 'MAINNET'
-                ? 'function inflation_rate() view returns (uint256)'
-                : 'function inflation_rate(uint256 week) view returns (uint256)';
-        const multicall = new Multicaller3([childChainAbi]);
-
-        let response: { [gaugeAddress: string]: ChildChainInfo } = {};
-
-        gaugeAddresses.forEach((address) => {
-            // Only L2 gauges have the inflation_rate with a week parameter
-            if (networkContext.chain === 'MAINNET') {
-                multicall.call(address, address, 'inflation_rate', [], true);
-            } else {
-                multicall.call(address, address, 'inflation_rate', [currentWeek], true);
+        for (const gauge of gauges) {
+            if (gauge.version === 2) {
+                this.balMulticaller.call(`${gauge.id}.rate`, gauge.id, 'inflation_rate', [currentWeek], true);
+                this.balMulticaller.call(`${gauge.id}.workingSupply`, gauge.id, 'working_supply', [], true);
+            } else if (networkContext.chain === Chain.MAINNET) {
+                this.balMulticaller.call(`${gauge.id}.weight`, networkContext.data.gaugeControllerAddress!, 'gauge_relative_weight', [gauge.id], true);
+                this.balMulticaller.call(`${gauge.id}.workingSupply`, gauge.id, 'working_supply', [], true);
             }
-        });
+        }
+        const balData = await this.balMulticaller.execute() as GaugeBalDistributionData;
 
-        const childChainData = (await multicall.execute()) as Record<string, string | undefined>;
-
-        for (const childChainGauge in childChainData) {
-            if (childChainData[childChainGauge]) {
-                response[childChainGauge] = {
-                    version: 2,
-                    rate: formatUnits(childChainData[childChainGauge]!, 18),
-                };
-            } else {
-                response[childChainGauge] = {
-                    version: 1,
-                    rate: '0.0',
-                };
+        // Get onchain data for reward tokens
+        for (const gauge of gauges) {
+            for (const token of gauge.tokens ?? []) {
+                const [address] = token.id.toLowerCase().split('-');
+                if (gauge.version === 1) {
+                    this.rewardsMulticallerV1.call(`${gauge.id}.rewardData.${address}`, gauge.id, 'reward_data', [address], true);
+                } else {
+                    this.rewardsMulticallerV2.call(`${gauge.id}.rewardData.${address}`, gauge.id, 'reward_data', [address], true);
```

Comment:
why the differentiation? It looks exactly the same.

Reply:
The return values from the reward_data function vary between the old and new L2 gauges.